### PR TITLE
feat(unstable): repurpose `--unstable-detect-cjs` to attempt loading more modules as cjs

### PR DESCRIPTION
**Pull request by dsherret**
This resurrects the `--unstable-detect-cjs` flag (which became stable), and repurposes it to attempt loading .js/.jsx/.ts/.tsx files as CJS in the following additional scenarios:

1. There is no package.json
1. There is a package.json without a "type" field

Also cleans up the implementation of this in the LSP a lot by hanging `resolution_mode()` off `Document` (didn't think about doing that until now).

Related issues:

* https://github.com/denoland/deno/issues/27079
* https://github.com/denoland/deno/issues/27056

### CODE REVIEW

**Comment by unimonkiez** [comment]
I actually much prefer using a package.json with type commonjs than this flag but that doesn't seem to work

File: cli/args/flags.rs [hunk 1]
```diff
@@ -4373,7 +4374,7 @@ impl CommandExt for Command {
     ).arg(
       Arg::new("unstable-detect-cjs")
         .long("unstable-detect-cjs")
-        .help("Reads the package.json type field in a project to treat .js files as .cjs")
+        .help("Treats ambiguous .js, .jsx, .ts, .tsx files as CommonJS modules in more cases")
         .value_parser(FalseyValueParser::new())
         .action(ArgAction::SetTrue)
         .hide(true)
```
Comment (bartlomieju) [inline cli/args/flags.rs:4380]: Unhide?

**Review by bartlomieju — APPROVED** [review]

File: cli/args/flags.rs (hunk 1 above)
Comment (dsherret) [reply cli/args/flags.rs:4380]: I thought so too, but all of the unstable flags seem hidden

Comment (bartlomieju) [reply cli/args/flags.rs:4380]: Oh, just make sure it's shown when you do `deno run --help=unstable`

**Review by nayeemrmn — APPROVED** [review]
LGTM

**Pull request merged**